test/e2e/reconciler/ingress/controller_test.go

```diff
@@ -68,55 +66,59 @@ func TestIngressController(t *testing.T) {
 		{
 			name: "ingress lifecycle",
 			work: func(ctx context.Context, t *testing.T, servers map[string]runningServer) {
+				// We create a root ingress. Ingress is excluded (through a hack) in namespace controller to be labelled.
+				// The ingress-controller will take over the labelling of the leaves. After that the normal syncer will
+				// sync the leaves into the physical cluster.
+
+				t.Logf("Creating ingress in source cluster")
 				ingressYaml, err := embeddedResources.ReadFile("ingress.yaml")
 				require.NoError(t, err, "failed to read ingress")
-
-				var ingress *v1.Ingress
-				err = yaml.Unmarshal(ingressYaml, &ingress)
-				require.NoError(t, err, "failed to create ingress")
-
-				ingress, err = servers[sourceClusterName].client.Ingresses(testNamespace).Create(ctx, ingress, metav1.CreateOptions{})
+				var rootIngress *v1.Ingress
+				err = yaml.Unmarshal(ingressYaml, &rootIngress)
+				require.NoError(t, err, "failed to unmarshal ingress")
+				rootIngress, err = servers[sourceServerName].client.Ingresses(testNamespace).Create(ctx, rootIngress, metav1.CreateOptions{})
 				require.NoError(t, err, "failed to create ingress")
 
-				nsLocator := syncer.NamespaceLocator{LogicalCluster: ingress.ClusterName, Namespace: ingress.Namespace}
+				nsLocator := syncer.NamespaceLocator{LogicalCluster: rootIngress.ClusterName, Namespace: rootIngress.Namespace}
 				targetNamespace, err := syncer.PhysicalClusterNamespaceName(nsLocator)
 				require.NoError(t, err, "error determining namespace mapping for %v", nsLocator)
 
-				expectedIngress := ingress.DeepCopy()
-				expectedIngress.GenerateName = ingress.Name + "-"
-				expectedIngress.Namespace = targetNamespace
-				err = servers[sinkClusterName].expect(expectedIngress, func(object *v1.Ingress) error {
-					if diff := cmp.Diff(expectedIngress.Spec, object.Spec); diff != "" {
-						return fmt.Errorf("saw incorrect spec on sink cluster: %s", diff)
+				t.Logf("Waiting for ingress to be synced to sink cluster to namespace %s", targetNamespace)
```

Comment:
am not a fan of source/dest because it sounds symmetric which kcp workspace and pcluster aren't. Upstream, downstream is better, source and sink too.